### ISSUE TITLE
allow CONSTRUCT subqueries in FROM clauses

### ISSUE THREAD

**Issue by joernhees** [w3c/sparql-12#33]
Why?
-----
This adds a lot of flexibility for example to do ad-hoc partitioning or (partial) alignments of graphs using different vocabularies.

Example:
```
g1:
g1:a skos:prefLabel "a"@en .
g1:a skos:prefLabel "A"@de .
g1:b rdfs:label "b" .

g2:
g2:A rdfs:label "A2" .

Query:
SELECT ?s ?al
FROM {
 CONSTRUCT {
  ?u :anyLabel ?l .
 }
 FROM <g1>
 FROM <g2> 
 WHERE {
  ?x ?p ?rl .
  VALUES ?p { skos:prefLabel rdfs:label }
  BIND (STR(?rl) AS ?l)
  BIND (IRI(REPLACE(STR(?x), "g2", "g1")) AS ?u)
 }
}
WHERE {
 ?s :anyLabel ?al .
}

Result:
?s    ?al
g1:a  "a"
g1:a  "A"
g1:a  "A2"
g1:b  "b"
```

Previous work
--------------
None that i'm aware of.

Considerations for backward compatibility
------------------------------------------
Extension only.

**Comment by jindrichmynarz** [comment]
Is this something that cannot be achieved in SPARQL 1.1, or do you propose it as a more convenient syntax?

I'm probably missing something. Given the data:

```trig
@base         <http://example.com/> .
@prefix g1:   <g1/> .
@prefix g2:   <g2/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .

<g1> {
  g1:a skos:prefLabel "a"@en .
  g1:a skos:prefLabel "A"@de .
  g1:b rdfs:label "b" .
}

<g2> {
  g2:A rdfs:label "A2" .
}
```
And a standard SPARQL 1.1 query:

```rq
BASE         <http://example.com/>
PREFIX g1:   <g1/>
PREFIX g2:   <g2/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?s ?al
FROM <g1>
FROM <g2> 
WHERE {
  VALUES ?p {
    rdfs:label
    skos:prefLabel
  }
  ?x ?p ?rl .
  BIND (str(?rl) AS ?al)
  BIND (iri(replace(str(?x), "g2", "g1")) AS ?s)
}
```

We get the result you want. What have I misunderstood?

**Comment by kasei** [comment]
@jindrichmynarz I'm not sure if this actually increases expressivity (I believe it does), but I've run across use cases for which I think this proposal would have been useful and the only solution in SPARQL 1.1 is either to perform multiple chained queries where one result is used to create the next query, or to try to put everything into a single query which results in a combinatorial explosion of the query size.

As a quick example, say you had a fairly complex query (q1) that resulted in a list of IRIs which represent named graphs, and you wanted to then perform another graph pattern match (q2) on the graph union of those named graphs. I believe this is possible in SPARQL 1.1, but only for q2 that don't involve property paths, and you will still end up with a *very* large query string which may grow beyond practical limits for using HTTP GET on an endpoint.

**Comment by joernhees** [comment]
@jindrichmynarz my example might be a bit simplistic, i agree. My main motivation is to be able to shift "data preparation" complexity out of the main query. I've had use-cases, where this would've been very beneficial, pretty much as @kasei points out: doing the mapping in a single spot rather than weaving it into the main query would've at least prettified the query a lot ;)

**Comment by tayloj** [comment]
I think I remember Rob Vesse saying that rdf. Net might have had something like this. The ability to construct from get you a lot of power that otherwise requires constructing into a model and then selecting from that model. A lot of that comes when the constructs from pattern uses filters or property paths. For instance you can't construct a concise bounded graph (CBG) in Sparql alone, but with select from construct, you can. See, for instance:

* https://stackoverflow.com/questions/9780385/how-to-recursively-expand-blank-nodes-in-sparql-construct-query#comment46516026_9780385 ( unfortunately, answers. Semantic web seems to be down, so the link in that comment is dead)

* https://stackoverflow.com/questions/17363199/how-to-get-a-concise-bounded-description-of-a-resource-with-sesame

( commenting from the phone, I may clean this up later)

**Comment by joernhees** [comment]
Cool example about path queries with restrictions, but I'd rename `ex:isBlank` to `ex:nonBlank` and also remove the mention of CBDs (and "Using that, it's possible to create a CBD"), as it's not that easy:

```
:a :p _:b1 .
_:b1 :q _:b2 .
_:b2 :q :c .
```

CBD of `:a` is all of the above.

**Comment by tayloj** [comment]
@joernhees Good catch on the naming (fixed) and the CBDs.  And you're right, it's not *quite* that easy (so I removed the final sentence, which was misleading).  But using this kind of construction, it's possible to get a CBD.  For the CBD, we want to follow every path from the root until we reach a *non-blank* node.  We need to know which nodes are blank and which are not, so something like:

```sql
construct {
  ?blank ex:blank ?blank ; ?p ?o .
  ?iri ex:blank ?iri ; ?pp ?oo .
where {
  { ?blank ?p ?o . filter isBlank(?blank) }
  union 
  { ?iri ?pp ?oo . filter (!isBlank(?iri)) }
}
```

Then if we want the CBD of `:a`, we want to follow every path from `:a` through blank nodes until we reach a non-blank node (and 

```sql
select ?s ?p ?o from /* construct as above */ where {
  :a ((<>|!<>)/ex:isBlank)* ?s . # includes :a as well
  ?s ?p ?o .
}
```

I think that would produce all of:

> ```
> :a :p _:b1 .
> _:b1 :q _:b2 .
> _:b2 :q :c .
> ```

I wish answers.semanticweb was still up.  This wasn't a trivial problem, but I'm pretty sure this feature makes it possible.

**Comment by TallTed** [comment]
@tayloj  - 
> https://stackoverflow.com/questions/9780385/how-to-recursively-expand-blank-nodes-in-sparql-construct-query#comment46516026_9780385 (unfortunately, answers.semanticweb seems to be down, so the link in that comment is dead, does anyone have a cache?

archive.org is our friend. There were 2 answers.semanticweb links in that SO comment --

* https://web.archive.org/web/20140722154003/http://answers.semanticweb.com/questions/26220/sparql-query-to-return-all-triples-recursively-that-make-up-an-rdfsclass-definition

* https://web.archive.org/web/20130121220310/http://answers.semanticweb.com/questions/20361/implementing-concise-bounded-description-in-sparql
